### ISSUE TITLE
Feature Request: RNN optimization when bucketing

### ISSUE THREAD

**Issue by ivodopyanov** [keras-team/keras#8337]
Hello
When dealing with variable length sequences we set some maximum sequence length and then pad all smaller sequences with 0. 
If we use bucketing (build batches from samples with approx. same length) then mask of batch would have rectangular 0 matrix at the end of size <size of batch>*<N>. These N steps do not affect the result and RNN computation could be completed before them (using condition in theano scan function or tensorflow while_loop). This could give big performance boost during training.

UPD: Probably you could do this trick if return_sequences=False: 
1) sum mask along all axis except timesteps
2) drop out steps from input where mask sum is 0
That would reduce number of loops.
